### PR TITLE
[expo-contacts][android] fix creating contacts with birthday

### LINKED ISSUE #30131: [expo-contact][Android] Creating a contact with a birthday does not work
### Minimal reproducible example

see summary example

### What platform(s) does this occur on?

Android

### Did you reproduce this issue in a development build?

Yes

### Summary

When you want to create a contact with a date, it doesn't work (the contact isn't created). If you remove the date in the Contact properties, it works fine.

Reading the code, dates are not taken into account with `getInsertOperation`.

For example:
```
import type { Contact} from 'expo-contacts';
import {
  addContactAsync,
} from 'expo-contacts';
...
const contact: Contact = {
    id: "ID",
    contactType: 'person',
    firstName: firstName ?? '',
    lastName: lastName ?? '',
    name: `${firstName ?? ''} ${lastName ?? ''}`,
    addresses: addresses.map(address => ({
      label: address[0],
      street: address[1],
      isPrimary: address[0] === 'Main',
      id: `${profileId}-${address[1]}`,
    })),
    phoneNumbers: phoneNumbers.map(phone => ({
      label: phone[0],
      number: phone[1],
      isPrimary: phone[0] === 'Main',
      id: `${profileId}-${phone[1]}`,
    })),
    emails: emails.map(email => ({
      label: email[0],
      email: email[1],
      isPrimary: email[0] === 'Main',
      id: `${profileId}-${email[1]}`,
    })),
    dates: birthdayDate
      ? [
          {
            label: 'birthday',
            year: birthdayDate?.getFullYear(),
            month: birthdayDate?.getMonth(),
            day: birthdayDate?.getDate(),
            id: `${profileId}-birthday`,
          },
        ]
      : [],
    urlAddresses: (userName
      ? [
          {
            label: 'azzapp',
            url: buildUserUrl(userName),
            id: `${profileId}-azzapp`,
          },
        ]
      : []
    )
  };
...
await addContactAsync(contact);
```

### Environment

```
  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.11.1 - ~/.nvm/versions/node/v20.11.1/bin/node
      Yarn: 3.8.2 - ~/.nvm/versions/node/v20.11.1/bin/yarn
      npm: 10.5.0 - ~/.nvm/versions/node/v20.11.1/bin/npm
      Watchman: 2024.01.22.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.15.2 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.4, iOS 17.4, macOS 14.4, tvOS 17.4, visionOS 1.1, watchOS 10.4
      Android SDK:
        API Levels: 31, 32, 33, 34
        Build Tools: 33.0.1, 34.0.0
        System Images: android-34 | Google APIs ARM 64 v8a, android-34 | Google Play ARM 64 v8a, android-VanillaIceCream | Google Play ARM 64 v8a
    IDEs:
      Android Studio: 2023.2 AI-232.10227.8.2321.11479570
      Xcode: 15.3/15E204a - /usr/bin/xcodebuild
    Expo Workflow: bare
```

### Expo Doctor Diagnostics

```
Detailed check results:

Expected package @expo/config-plugins@~8.0.0
Found invalid:
  @expo/config-plugins@7.8.4
  (for more info, run: npm why @expo/config-plugins)
Expected package metro@~0.80.8
Found invalid:
  metro@0.80.5
  (for more info, run: npm why metro)
Expected package metro-resolver@~0.80.8
Found invalid:
  metro-resolver@0.80.5
  (for more info, run: npm why metro-resolver)
Expected package metro-config@~0.80.8
Found invalid:
  metro-config@0.80.5
```

### PR DESCRIPTION
**Pull request by Eric013**
# Why

Fix #30131 

# How

I use expo-contacts to create contacts with birthday dates and I've noticed that when I specify a date in my Contact it doesn't work (not taken into account or the contact isn't created).

# Test Plan

By using addContactAsync with a contact of type Contact with and without a date property attached to the contact

```
import type { Contact} from 'expo-contacts';
import {
  addContactAsync,
} from 'expo-contacts';
...
const contact: Contact = {
    id: "ID",
    contactType: 'person',
    firstName: firstName ?? '',
    lastName: lastName ?? '',
    ...
    name: `${firstName ?? ''} ${lastName ?? ''}`,
    dates: birthdayDate
      ? [
          {
            label: 'birthday',
            year: birthdayDate?.getFullYear(),
            month: birthdayDate?.getMonth(),
            day: birthdayDate?.getDate(),
            id: `${profileId}-birthday`,
          },
        ]
      : [],
    urlAddresses: (userName
      ? [
          {
            label: 'azzapp',
            url: buildUserUrl(userName),
            id: `${profileId}-azzapp`,
          },
        ]
      : []
    )
  };
...
await addContactAsync(contact);
```

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by vonovak — CHANGES_REQUESTED** [review]
Thank you for the PR.

With this change, the code does more work than necessary afaict:

In `fromMap` it takes individual map entries to convert them into string in `parseDateFromMap`.
Then it takes the string and parses it back into the individual numbers using `SimpleDateFormat.parse`. It can just take the input directly, no?

Also, note that looking at the code, it seems that `year` is optional. Maybe other fields are as well. Let's try to make the code work without the `year` field being provided.

Finally, would you mind creating a [test case here](https://github.com/expo/expo/blob/main/apps/test-suite/tests/Contacts.js)?

Thank you!

**Comment by mlecoq** [comment]
> Thank you for the PR.
> 
> With this change, the code does more work than necessary afaict:
> 
> In `fromMap` it takes individual map entries to convert them into string in `parseDateFromMap`. Then it takes the string and parses it back into the individual numbers using `SimpleDateFormat.parse`. It can just take the input directly, no?
> 
> Also, note that looking at the code, it seems that `year` is optional. Maybe other fields are as well. Let's try to make the code work without the `year` field being provided.
> 
> Finally, would you mind creating a [test case here](https://github.com/expo/expo/blob/main/apps/test-suite/tests/Contacts.js)?
> 
> Thank you!

Hi @vonovak, following your suggestions I have made some updates. 

Thanks,
Mickael

File: packages/expo-contacts/android/src/main/java/expo/modules/contacts/Contact.kt [hunk 1]
```diff
@@ -424,8 +422,10 @@ class Contact(var contactId: String) {
         try {
           hasYear = !dateString!!.startsWith("--")
           if (hasYear) {
+            val datePattern = SimpleDateFormat("yyyy-MM-dd", Locale.US)
```
Comment (vonovak) [inline packages/expo-contacts/android/src/main/java/expo/modules/contacts/Contact.kt:425]: nothing wrong with `Locale.getDefault()` in this case, but I pinned it to `Locale.US` to make it "raise no questions" if that makes sense.

**Review by vonovak — APPROVED** [review]

**Review by lukmccall — APPROVED** [review]
LGTM

**Comment by vonovak** [comment]
@Eric013 thank you for contributing and thank you for your patience with review! 👍

**Pull request merged**